oak/server/rust/oak_runtime/src/runtime/mod.rs

```diff
@@ -169,25 +187,124 @@ impl Runtime {
+        } else {
+            error!("filter_handle: node {:?} not found", node_id);
+            None
+        }
+    }
+
+    /// Replace [`Handle`]s with [`None`] if the [`NodeId`] does not have access to the [`Handle`]s.
+    /// Any iterator is accepted, but a Vec<Option<Handle>> is returned so as not to hold
+    /// underlying Mutexes open.
+    fn filter_optional_handles<'a, I: Iterator<Item = &'a Option<Handle>>>(
+        &self,
+        node_id: NodeId,
+        handles: I,
+    ) -> Vec<Option<Handle>> {
+        // Allow RUNTIME_NODE_ID access to all handles.
+        if node_id == RUNTIME_NODE_ID {
+            return handles.copied().collect();
+        }
+
+        let nodes = self.nodes.read().unwrap();
+        // Lookup the node_id in the runtime's nodes hashmap
+        if let Some(node) = nodes.get(&node_id) {
+            let tracked_handles = node.handles.lock().unwrap();
+
+            // Map the handles iterator which contains Option<Handle>
+            handles
+                .map(|optional_handle| {
+                    // Map only Some(handle)s, leaving 'None's inplace
+                    optional_handle.and_then(|handle| {
+                        // Check the handle exists in the handless associated with a node, otherwise
+                        // return None.
+                        if tracked_handles.contains(&handle) {
+                            Some(handle)
+                        } else {
+                            error!(
+                                "filter_optional_handles: handle {:?} not found in node {:?}",
+                                handle, node_id
+                            );
+                            None
+                        }
+                    })
+                })
+                .collect()
+        } else {
+            error!("filter_optional_handles: node {:?} not found", node_id);
+            handles.map(|_| None).collect()
         }
     }
 
     /// Creates a new [`Channel`] and returns a `(writer handle, reader handle)` pair.
-    pub fn new_channel(&self, label: &oak_abi::label::Label) -> (Handle, Handle) {
-        self.channels.new_channel(label)
+    pub fn new_channel(&self, node_id: NodeId, label: &oak_abi::label::Label) -> (Handle, Handle) {
+        let (writer, reader) = self.channels.new_channel(label);
+        self.track_handles_in_node(node_id, vec![writer, reader].iter());
+        (writer, reader)
     }
 
     /// Reads the statuses from a slice of `Option<&ChannelReader>`s.
     /// [`ChannelReadStatus::InvalidChannel`] is set for `None` readers in the slice. For `Some(_)`
     /// readers, the result is set from a call to `has_message`.
-    fn readers_statuses(&self, readers: &[Option<Handle>]) -> Vec<ChannelReadStatus> {
+    fn readers_statuses(
+        &self,
+        node_id: NodeId,
+        readers: &[Option<Handle>],
+    ) -> Vec<ChannelReadStatus> {
+        let readers = self.filter_optional_handles(node_id, readers.iter());
```

Comment:
I think this is too lenient, why not just error out immediately if any handle is not accessible by this node? That suggests something bad is going on, right?

Reply:
I've changed filter_optional_handles to validate_handles_access and error immediately in `wait_on_channels`, i've removed the check here as this method is non-public and only called by  `wait_on_channels`